src/composites.rs

```diff
@@ -0,0 +1,214 @@
+//! Module for combining hyper services
+
+use std::{io, fmt};
+use hyper::server::{Service, NewService};
+use hyper::{Request, Response, StatusCode};
+use futures::{future, Future};
+use context::Context;
+
+/// Trait for getting the path of a request
+pub trait HasPath {
+    /// Retrieve the path
+    fn path(&self) -> &str;
+}
+
+impl HasPath for Request {
+    fn path(&self) -> &str {
+        self.path()
+    }
+}
+
+impl HasPath for (Request, Context) {
+    fn path(&self) -> &str {
+        self.0.path()
+    }
+}
+
+/// Trait for generating a default "not found" response
+pub trait HasNotFound {
+    /// Return a "not found" response
+    fn not_found() -> Self;
+}
+
+impl HasNotFound for Response {
+    fn not_found() -> Self {
+        Response::new().with_status(StatusCode::NotFound)
+    }
+}
+
+/// Trait for wrapping hyper NewServices to uniformize the return type of new_service().
+/// This is necessary in order for the NewServices with different Instance types to
+/// be stored in a single collection.
+pub trait BoxedNewService<U, V, W> {
+    /// Create a new Service trait object
+    fn boxed_new_service(
+        &self,
+    ) -> Result<
+        Box<
+            Service<
+                Request = U,
+                Response = V,
+                Error = W,
+                Future = Box<Future<Item = V, Error = W>>,
+            >,
+        >,
+        io::Error,
+    >;
+}
+
+impl<T, U, V, W> BoxedNewService<U, V, W> for T
+where
+    T: NewService<Request = U, Response = V, Error = W>,
+    T::Instance: Service<Future = Box<Future<Item = V, Error = W>>>
+        + 'static,
+{
+    /// Call the new_service() method of the wrapped NewService and Box the result
+    fn boxed_new_service(
+        &self,
+    ) -> Result<
+        Box<
+            Service<
+                Request = U,
+                Response = V,
+                Error = W,
+                Future = Box<Future<Item = V, Error = W>>,
+            >,
+        >,
+        io::Error,
+    > {
+        let service = self.new_service()?;
+        Ok(Box::new(service))
+    }
+}
+
+/// A struct combining multiple NewServices. The NewServices are stored in a list
+/// in the order they are added, together with an associated base path. The service
+/// generated by the CompositeNewService will pass a request to the first service
+/// in the list whose base path is a prefix of the request's path or return a
+/// "not found" response if there is no match.
+pub struct CompositeNewService<U, V, W>(Vec<(&'static str, Box<BoxedNewService<U, V, W>>)>)
+where
+    U: HasPath,
+    V: HasNotFound + 'static,
```

Comment:
I'm a bit surprised that this needs to be static. Any idea why?

Reply:
I think it's because this is put into the future that `call` returns, so it needs to live at lease as long as that future. Since we don't know how long that is, it needs to be static.